texk/web2c/man/cweb.man
.\" Automatically generated by Pandoc
.\"
.TH "CWEB" "1" "April 30, 2025." "Web2c @VERSION@" "General Commands Manual"
.SH NAME
ctangle, cweave \- translate CWEB to C/C++ and/or TeX
.SH SYNOPSIS
\f[B]ctangle\f[R] [\f[I]options\f[R]] \f[I]webfile\f[R][.w]
[{\f[I]changefile\f[R][.ch]|\-} [\f[I]outfile\f[R][.c]]]
.PD 0
.P
.PD
\f[B]cweave\f[R] [\f[I]options\f[R]] \f[I]webfile\f[R][.w]
[{\f[I]changefile\f[R][.ch]|\-} [\f[I]outfile\f[R][.tex]]]
.SH DESCRIPTION
The \f[B]ctangle\f[R] program converts a CWEB source document into a
C/C++\ program that may be compiled in the usual way.
The output file includes \f[B]#line\f[R] specifications so that
debugging can be done in terms of the CWEB source file.
.PP
The \f[B]cweave\f[R] program converts the same CWEB file into a
TeX\ file that may be formatted and printed in the usual way.
It takes appropriate care of typographic details like page layout and
the use of indentation, \f[I]italics\f[R], \f[B]boldface\f[R], etc., and
it supplies extensive cross-index information that it gathers
automatically.
.PP
CWEB allows you to prepare a single document containing all the
information that is needed both to produce a compilable C/C++\ program
and to produce a well-formatted document describing the program in as
much detail as the writer may desire.
The user of CWEB ought to be familiar with TeX as well as C/C++.
.SH USAGE
The command line should have one, two, or three names on it.
The first is taken as the CWEB input file (and \f[B].w\f[R] is added if
there is no extension).
If there is a second name, it is a change file (and \f[B].ch\f[R] is
added if there is no extension).
The change file overrides parts of the CWEB file, as described in the
documentation.
If there is a third name, it overrides the default name of the output
file, which is ordinarily the same as the name of the input file (but on
the current directory) with the extension \f[B].tex\f[R].
If you just want to change the output file name, but don\(cqt have a
change file to apply, you can use `\f[B]\-\f[R]' as the second argument.
.SH DIFFERENCES TO ORIGINAL CWEB
CWEBbin tries hard to be a drop-in replacement for CWEB, so in general
you should not notice any differences in invoking the programs nor in
the resulting output.
There are, however, a few differences worth noting:
.IP \(bu 2
Options \f[B]\-\-help\f[R], \f[B]\-\-quiet\f[R], \f[B]\-\-verbose\f[R],
\f[B]\-\-version\f[R], and flags \f[B]+c\f[R], \f[B]\-i\f[R],
\f[B]\-o\f[R], \f[B]+u\f[R], and \f[B]+lX\f[R] are new in CWEBbin and
TeX\ Live.
.IP \(bu 2
Option \f[B]+lX\f[R] is accompanied by several wrapper files for
\f[B]cwebmac.tex\f[R] with translated captions for German
(\f[B]+ld\f[R]), French (\f[B]+lf\f[R]), and Italian (\f[B]+li\f[R]).
.IP \(bu 2
File lookup with the environment variable CWEBINPUTS is extended to
permit several, colon-separated, paths; see ENVIRONMENT below.
.IP \(bu 2
If properly configured, the main programs \f[B]ctangle\f[R] and
\f[B]cweave\f[R] are localized with the \(lqGNU gettext utilities\(rq.
.SH OPTIONS
Options on the command line may be either turned off with `\f[B]\-\f[R]'
(if they are on by default) or turned on with `\f[B]+\f[R]' (if they are
off by default).
In fact, the options are processed from left to right, so a sequence
like \f[B]\-\-verbose \-h\f[R] will only show the \f[B]banner line\f[R]
(\f[B]+b\f[R]) and the \f[B]progress report\f[R] (\f[B]+p\f[R]), but
leave out the \f[B]happy message\f[R] (\f[B]\-h\f[R]).
.PP
The first batch of options are common to both \f[B]ctangle\f[R] and
\f[B]cweave\f[R]:
.IP \(bu 2
\f[B]+b\f[R]: print banner line on terminal
.IP \(bu 2
\f[B]+h\f[R]: print success message on completion
.IP \(bu 2
\f[B]+p\f[R]: print progress report messages
.IP \(bu 2
\f[B]+q\f[R]/\f[B]\-q\f[R]: shortcut for \f[B]\-bhp\f[R]; also
\f[B]\-\-quiet\f[R]
.IP \(bu 2
\f[B]+v\f[R]/\f[B]\-v\f[R]: shortcut for \f[B]+bhp\f[R]; also
\f[B]\-\-verbose\f[R] (default)
.IP \(bu 2
\f[B]+c\f[R]: check temporary output for changes
.IP \(bu 2
\f[B]+s\f[R]: print usage statistics
.IP \(bu 2
\f[B]\-\-help\f[R]: display help message and exit
.IP \(bu 2
\f[B]\-\-version\f[R]: output version information and exit
.PP
There are two other options applicable to \f[B]ctangle\f[R] only:
.IP \(bu 2
\f[B]+k\f[R]: keep separators in numeric literals in the output
.IP \(bu 2
\f[B]+u\f[R]: transliterate UTF-8 characters in C code
.PP
There are eight other options applicable to \f[B]cweave\f[R] only:
.IP \(bu 2
\f[B]\-e\f[R]: do not enclose C/C++\ material in \f[B]\(rsPB{\&...}\f[R]
.IP \(bu 2
\f[B]\-f\f[R]: do not force a newline after every C/C++\ statement in
output
.IP \(bu 2
\f[B]\-F\f[R]: do not force a compound statement to start on a new line
.IP \(bu 2
\f[B]\-i\f[R]: suppress indentation of parameter declarations
.IP \(bu 2
\f[B]\-o\f[R]: suppress separation of declarations and statements
.IP \(bu 2
\f[B]\-x\f[R]: omit indices, section names, table of contents
.IP \(bu 2
\f[B]+lX\f[R]/\f[B]\-lX\f[R]: use macros for language \f[I]X\f[R] as of
\f[I]X\f[R]\f[B]cwebmac.tex\f[R]
.IP \(bu 2
\f[B]+t\f[R]: treat \f[B]typename\f[R] in a template like
\f[B]typedef\f[R]
.SH ENVIRONMENT
The environment variable CWEBINPUTS is used to search for the input
files, or the system default if CWEBINPUTS is not set.
See tex(1) for the details of the searching.
To avoid conflicts with other programs that also use the CWEBINPUTS
environment, you can be more specific and use CWEBINPUTS_cweb for
special requirements in CWEB.
.PP
If prepared for NLS support, \f[B]ctangle\f[R] and \f[B]cweave\f[R] use
the environment variable TEXMFLOCALEDIR to configure the parent
directory where the \(lqGNU gettext utilities\(rq search for translation
catalogs.
.PP
These variables are preconfigured in TeX\ Live\(cqs \f[B]texmf.cnf\f[R].
.SH FILES
The location of the files mentioned below varies from system to system.
Use the \f[B]kpsewhich\f[R] utility to find their locations.
.IP \(bu 2
\f[B]cwebmac.tex\f[R]: The default TeX\ macros \f[B]\(rsinput\f[R] in
the first line of the \f[B]cweave\f[R] output file.
.IP \(bu 2
\f[B]cwebman.tex\f[R]: The CWEB user manual, available in PDF from \c
.UR https://ctan.org/pkg/cweb
CTAN
.UE \c
\&.
.SH SEE ALSO
.IP \(bu 2
The CWEB System of Structured Documentation: by Donald E.\ Knuth and
Silvio Levy (hardcopy version of \f[B]cwebman.tex\f[R] and the source
code listings of \f[B]common.w\f[R], \f[B]ctangle.w\f[R], and
\f[B]cweave.w\f[R]).
.IP \(bu 2
Literate Programming: by D.\ E.\ Knuth.
.IP \(bu 2
Weaving a Program: by Wayne Sewell.
.PP
cweb(1), tex(1), cc(1)
.SH AUTHORS
Don Knuth wrote WEB for TeX and Pascal.
.PD 0
.P
.PD
Silvio Levy designed and developed CWEB by adapting the WEB conventions
to C and by recoding everything in CWEB.
Knuth began using CWEB and made further refinements.
.PD 0
.P
.PD
Many other helpers are acknowledged in the CWEB manual.
.PD 0
.P
.PD
Contemporary development on https://github.com/ascherer/cweb.


ChangeLog
2025-07-22  Karl Berry  <karl@freefriends.org>

	* reautoconf: mention that autoconf-only packages need to be
	explicitly added; would be better to read kpse-pkgs.m4.
	Also reautoconf tree for kpse-setup.m4 changes.

2025-06-26  Karl Berry  <karl@freefriends.org>

	* automake 1.18.1.

2025-04-30  Karl Berry  <karl@freefriends.org>

	* Build: incorporate original CFLAGS, CXXFLAGS, OBJCXXFLAGS values.
	https://tug.org/pipermail/tlbuild/2025q2/005771.html

2025-04-29  Karl Berry  <karl@freefriends.org>

	* Build: use TL_CFLAGS, TL_CXXFLAGS, etc., for TL_BUILD_ENV
	if -g is not given.

2025-03-08  Karl Berry  <karl@tug.org>

	* version.ac: 2026/dev.

2025-03-07  Karl Berry  <karl@tug.org>

	* TL'25 release.

2025-01-12  Karl Berry  <karl@tug.org>

	* Build: also support --debug-more for -g -Og -ggdb3,
	per Luigi, as described on https://tug.org/texlive/build.html.

2024-07-12  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake 1.17.

2024-04-29  Karl Berry  <karl@freefriends.org>

	* Makefile.am (distdir): override default dist generation,
	since we don't try to keep make dist working any more.
	https://tug.org/pipermail/tlbuild/2024q2/005547.html

2023-12-31  Karl Berry  <karl@freefriends.org>

	* reautoconf: add -f / --force option to pass to autoreconf.

2023-12-26  Karl Berry  <karl@freefriends.org>

	* reautoconf: don't bother with environment variables;
	because the autotools have many subprograms, setting PATH
	is what makes sense.

2023-12-23  Karl Berry  <karl@tug.org>

	* reautoconf with autoconf-2.72.

2023-12-21  Karl Berry  <karl@freefriends.org>

	* reautoconf (do_it): abort the run if any command fails.

2023-02-25  Karl Berry  <karl@tug.org>

	* Build: show full uname -a.

2023-02-09  Karl Berry  <karl@freefriends.org>

	* version.ac: 2023, no more /dev.

2022-03-21  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): 2022-03-21 for TL'22.

2022-01-18  Karl Berry  <karl@freefriends.org>

	* version.ac: 2022, no more /dev.

2021-10-04  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake-1.16.5.

2021-06-27  Karl Berry  <karl@freefriends.org>

	* reautoconf: pass -Wno-obsolete by default, since Autoconf 2.70
	and later obsoletes many common macros. See m4/README for more.
	Also standardize shell usage, etc., but no other functional changes.

2021-04-17  Karl Berry  <karl@tug.org>

	* version.ac: now 2022/dev since TL'21 is released.

2021-03-23  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-23 for TL'21.

2021-02-17  Karl Berry  <karl@tug.org>

	* version.ac: 2021, no more /dev.

2020-05-28  Karl Berry  <karl@freefriends.org>

	* Makefile.am (world): do not descend into $(texlinks_dir) if it
	doesn't exist. Report from Andreas Scherer,
	https://tug.org/pipermail/tlbuild/2020q1/004614.html.

2020-05-14  Karl Berry  <karl@freefriends.org>

	* reautoconf for removal of poppler support.

2020-04-10  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2020-03-27  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-27 for TL'20.

2020-03-06  Karl Berry  <karl@tug.org>

	* .travis.yml: attempt to reenable musl build on github.

2020-01-31  Karl Berry  <karl@tug.org>

	* Makefile.am (texlinks_dir): now texk/texlive/tl_support
	(instead of tl_scripts).

2020-01-16  Karl Berry  <karl@freefriends.org>

	* version.ac (tex_live_version): remove /dev, preparing for TL'20.

2019-06-25  Karl Berry  <karl@freefriends.org>

	* Build: don't call buildinfo from here, the output when the CC
	etc. envvars are not set is too misleading. Just write UNAME.

2019-06-24  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh (compiler_version): new shell function;
	 ignore useless first "Configured:" line for gcc->cc on Macs.

2019-06-23  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: output the compiler/other names
	that we pass from the buildinfo target.
	* Makefile.am (buildinfo): pass OBJCXX, not OBJC.
	* Build: more maintainable exports of TL_*.

2019-06-22  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: rename from build-info.sh;
	forget /etc/issue;
	allow option to omit config.status output;
	do not dump environment;
	run lsb_release -a. We'll see ...
	* Build: use buildinfo.sh instead of inline commands (mostly).
	* Makefile.am (buildinfo): make dependency of all-recursive
	so that it runs first; rename from build-info; remove from world.

2019-06-06  Karl Berry  <karl@tug.org>

	* Makefile.am (world): make build-info.
	(build-info): new target to invoke doc/build-info.sh.
	(build_info_target): new variable.
	* Build: export TL_* variables so build-info.sh can report them.
	* doc/build-info.sh: new script to report compiler versions,
	dump environment, config.status, etc.

2019-05-03  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2019-04-07  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): let's hope for 2019-04-07.

2018-10-02  Karl Berry  <karl@tug.org>

	* Build: TL_MAKE_VERBOSE=VERBOSE=1 by default,
	so that test-suite.log will be shown on failure.
	https://lists.gnu.org/archive/html/automake/2018-09/msg00015.html

2018-04-28  Karl Berry  <karl@tug.org>

	* version.ac: 2019/dev.

2018-04-10  Karl Berry  <karl@tug.org>

	* README.*: update from new tlbuild.

2017-11-15  Karl Berry  <karl@tug.org>

	* *.test: use #!/bin/sh -vx. Add 

2017-03-21  Karl Berry  <karl@freefriends.org>

	* everywhere: use EXTRA_DIST=TLpatches instead of <pkgname>-PATCHES.

2017-03-18  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2017/dev => 2017 for TL17.

2016-11-16  Karl Berry  <karl@tug.org>

	* Makefile.am (install-exec-hook): warn that the
	result of make install is not in itself usable.

2016-06-30  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2016 => 2017/dev.
	* tardate.ac (tex_live_tardate): 2015-05-12 => 2016-05-23.
	(Though this is not used for any release.)

2016-03-16  Karl Berry  <karl@tug.org>


texk/web2c/cwebdir/cweb.el
;; This file contains extensions to GNU-Emacs, to wit:
; (1) some WEB-oriented functions that are also of general use
; (2) changes to the GNU-distributed TeX mode
; (3) definitions of simple WEB and CWEB modes

; To use: Put this in your EMACS-lisp library and say (load-library "cweb")
; in your .emacs init file.

; Contributed by Don Knuth, July 1990

;; OK, here's part (1): some WEB-oriented functions whose main purpose is
; to maintain a stack of module names that are "pending" as you are writing
; a program. When you first think of a module that needs to be written later,
; put it into the pending list (by typing CTL-Z instead of @> after the
; name). Later you can say CTL-\ to retrieve a pending name (and if
; you want to cycle through the pending names, ESC-y after CTL-\ will
; do it, just as ESC-y works after a yank).
; After you've said CTL-\, the current region is the name just removed from
; the pending list. If you change your mind, you can put it back again by
; saying ESC-\. If you had put it into the pending list by mistake, you
; can get rid of it by using the normal CTL-W operation (kill-region).
; The following code binds the new commands to CTL-Z, CTL-\, and ESC-\
; in all modes. You may prefer other bindings, of course.
; CTL-Z is normally "suspend emacs", but it is best not used when emacs
; has its own window as it usually does nowadays; if you need the
; old CTL-Z, you might rather bind it to CTL-X CTL-Z.
; CTL-\ is normally undefined.
; ESC-\ is normally "delete space", but ESC-space DEL does that easily too.

(defvar pending-list nil
 "List of strings (usually WEB module names) still pending.")

(defun into-pending-list (beg end)
 "Copy region into pending-list."
 (interactive "r")
 (indicate-region)
 (setq pending-list (cons (buffer-substring beg end) pending-list)))

(defun new-module-name-pending ()
 "Insert @> to complete a module name, then put it into pending-list."
 (interactive)
 (insert "@>")
 (push-mark)
 (if (search-backward "@<" nil t)
     (progn
       (exchange-point-and-mark)
       (into-pending-list (point) (mark))
       )
   (message "There's no @< to begin the module name!")))
(global-set-key "\C-z" 'new-module-name-pending)

(defun pop-pending-list (arg)
 "Remove first element of pending-list and insert it as current region.
With argument, put point at left; otherwise point will follow the insertion.
Say \\[new-yank-pop] to replace this by another element of the list.
Say \\[into-pending-list] to put it back in the list."
 (interactive "*P")
 (if (consp pending-list)
     (progn
       (push-mark (point))
       (insert (car pending-list))
       (setq pending-list (cdr pending-list))
       (if arg
           (exchange-point-and-mark)))
   (message "Nothing is pending.")
   (setq this-command nil)))
(global-set-key "\C-\\" 'pop-pending-list)
(global-set-key "\M-\\" 'into-pending-list)

(defun new-yank-pop (arg)
 "If previous command was \\[pop-pending-list], pop a different string;
otherwise do an ordinary Meta-y."
 (interactive "*p")
 (if (eq last-command 'pop-pending-list)
     (let (xch)
       (setq xch (< (point) (mark)))
       (setq pending-list (append pending-list
                                 (list (buffer-substring (point) (mark)))))
       (delete-region (point) (mark))
       (setq this-command 'pop-pending-list)
       (pop-pending-list xch))
   (yank-pop arg)))
(global-set-key "\M-y" 'new-yank-pop)

(defun indicate-region ()
  "Bounce cursor to mark and back again"
  (let ((point-save (point)))
    (unwind-protect
        (progn (goto-char (mark))
               (sit-for 0 300)) ;; wait 300 milliseconds
      (goto-char point-save))))

; I prefer to change the standard copy-region command to the following,
; which gives me visual feedback about what I've copied to the kill ring:
(defun indicate-and-copy-region (beg end)
  "Indicate current region, then copy it to the kill ring."
  (interactive "r")(indicate-region)(copy-region-as-kill beg end))
(global-set-key "\M-w" 'indicate-and-copy-region)

; Here's another convenient command, bound to the usually unused ESC-".
(defun ditto (arg)
  "Copy ARG characters from the line above."
  (interactive "*p")
  (let (ch)
    (while (> arg 0)
      (setq temporary-goal-column (current-column))
      (save-excursion
        (previous-line 1)
        (setq ch (following-char)))
      (insert ch)
      (setq arg (1- arg)))))
(global-set-key "\M-\"" 'ditto)
; If "ditto" suddenly fails to work, you have probably set goal-column
; inadvertently. To unset it, say C-u C-x C-n.

;; OK, here's part (2): Changes to TeX mode.
; The WEB modes below are very much like TeX mode, but some improvements were
; desirable in TeX mode:
; I made newline act as it does in indented-text mode, since this
; works nicely for both TeX and WEB (Pascal or C code).
; I made RET check for unmatched delimiters if it ends a paragraph.
; Otherwise TeX mode remains as it was before.

(setq TeX-mode-map (make-sparse-keymap))
(define-key TeX-mode-map "\C-c\C-k" 'TeX-kill-job)
(define-key TeX-mode-map "\C-c\C-l" 'TeX-recenter-output-buffer)
(define-key TeX-mode-map "\C-c\C-q" 'TeX-show-print-queue)
(define-key TeX-mode-map "\C-c\C-p" 'TeX-print)
(define-key TeX-mode-map "\"" 'TeX-insert-quote)
(define-key TeX-mode-map "\e}" 'up-list)
(define-key TeX-mode-map "\e{" 'TeX-insert-braces)
(define-key TeX-mode-map "\C-c\C-r" 'TeX-region)
(define-key TeX-mode-map "\C-c\C-b" 'TeX-buffer)
(define-key TeX-mode-map "\C-c\C-f" 'TeX-close-LaTeX-block)
(define-key TeX-mode-map "\r" 'TeX-newline)
(define-key TeX-mode-map "\t" 'indent-relative)
(setq TeX-mode-hook '(lambda ()
  (make-local-variable 'indent-line-function)
  (setq indent-line-function 'indent-relative-maybe)))

(defun TeX-newline (arg)
"If previous character is newline and no ARG, check for unbalanced braces
and/or dollar signs in previous paragraph. If ARG is \\[universal-argument],
do a single newline; otherwise do ordinary newline."
 (interactive "*P")
 (if (and (eq (preceding-char) ?\n) (not arg))
     (TeX-check-paragraph)
   (if (listp arg)
       (newline)
     (newline arg))))

(defun TeX-check-paragraph ()
"Insert a newline following a newline, breaking a paragraph for TeX.
Check for mismatched delimiters in paragraph being terminated."
  (interactive)
  (if (TeX-validate-paragraph
           (save-excursion
             (search-backward "\n\n" nil 'move)
             (point))
           (point))
      (insert ?\n)
    (insert ?\n)
    (error "Mismatched delimiters in that paragraph?")))

;; and now, part (3): WEB and CWEB modes.
; These are like plain TeX mode except that the automatic conversion of
; " to `` or '' is disabled. (Personally I never liked that feature anyway,
; since it's easy to get used to typing `` and ''. In WEB modes, the
; feature soon becomes intolerable, unless you never use string constants!)
; Another thing distinguishing WEB mode from TeX is ESC-p and ESC-n, to
; move to previous or next module. These keys are usually unbound, except
; when processing email.

(defun forward-module (arg)
"Advance past next WEB module beginning; with ARG, repeat ARG times."
 (interactive "p")
 (move-to-module arg))
(defun backward-module (arg)
"Advance to previous WEB module beginning; with ARG, repeat ARG times."
 (interactive "p")
 (move-to-module (- arg)))
(defun move-to-module (arg)
 (while (> arg 0)
   (re-search-forward "@ \\|@\\*\\|@\n")
   (setq arg (1- arg)))
 (while (< arg 0)
   (re-search-backward "@ \\|@\\*\\|@\n")
   (setq arg (1+ arg))))

(defun web-mode ()
  "Major mode like TeX mode plus \\[forward-module] and \\[backward-module]
for relative module movement. The automatic \" feature is disabled."
  (interactive)
  (plain-tex-mode)
  (local-set-key "\M-n" 'forward-module)
  (local-set-key "\M-p" 'backward-module)
  (local-set-key "\"" 'self-insert-command)
  (setq mode-name "WEB")
  (setq major-mode 'web-mode)
  (run-hooks 'web-mode-hook))
(setq auto-mode-alist (cons '("\\.web$" . web-mode) auto-mode-alist))

(defun cweb-mode ()
  "Major mode like TeX mode plus \\[forward-module] and \\[backward-module]
for relative module movement. The automatic \" feature is disabled."
  (interactive)
  (plain-tex-mode)
  (local-set-key "\M-n" 'forward-module)
  (local-set-key "\M-p" 'backward-module)
  (local-set-key "\"" 'self-insert-command)
  (setq comment-start nil)
  (modify-syntax-entry ?% "@")
  (setq mode-name "CWEB")
  (setq major-mode 'cweb-mode)
  (setq tex-fontify-script nil) ;; needed in GNU Emacs version 22?
  (run-hooks 'cweb-mode-hook))
(setq auto-mode-alist (cons '("\\.w$" . cweb-mode) auto-mode-alist))
(setq auto-mode-alist (cons '("\\.ch$" . cweb-mode) auto-mode-alist))
